proposals/debug-id.md

```diff
@@ -0,0 +1,237 @@
+
+## Scope
+
+This proposal sets some specific limitations on source maps to simplify the
+processing in the wider ecosystem.  Debug IDs are at present only specified to
+source maps with embedded sources or where sources are categorically not
+available.  The lookup for original sources from a source map identified by a
+debug ID is not defined.
+
+Additionally, this specification applies only to non-indexed source maps and
+currently specifies references only for JavaScript.
+
+## Terms
+
+In the context of this document:
+
+- **Source Map:** Refers to a non-indexed, standard source map.
+- **Transpiled File:** Refers to a transpiled (potentially minified) JavaScript file.
+- **Debug ID:** Refers to a UUID as described in this document.
+
+## Debug IDs
+
+Debug IDs are globally unique identifiers for build artifacts.  They are
+specified to be UUIDs.  In the context of this proposal, they are represented in
+hexadecimal characters.  When comparing debug IDs they must be normalized.  This
+means that `85314830-023f-4cf1-a267-535f4e37bb17` and
+`85314830023F4CF1A267535F4E37BB17` are equivalent but the former representation
+is the canonical format.
+
+The way a debug ID is generated is specific to the toolchain and no requirements
+are placed on it.  It is however recommended to generate deterministic debug IDs
+(UUID v3 or v5) so that rebuilding the same artifacts yields stable IDs.
+
+Debug IDs are embedded in both source maps and transpiled files, allowing a
+bidirectional mapping between them. The linking of source maps and transpiled
+files via HTTP headers is explicitly not desired.  A file identified by a Debug
+ID must have that Debug ID embedded to ensure the file is self-identifying.
+
+### Debug IDs in Source Maps
+
+We propose adding a `debugId` property to the source map at the top level of
+the source map object.  This property should be a string value representing
+the Debug ID in hexadecimal characters, preferably in the canonical UUID
+format:
+
+```json
+{
+  "version": 3,
+  "file": "app.min.js",
+  "debugId": "85314830-023f-4cf1-a267-535f4e37bb17",
+  "sources": [...],
+  "sourcesContent": [...],
+  "mappings": "..."
+}
+```
+
+### Debug IDs in JavaScript Artifacts
+
+Transpiled JavaScript files containing a Debug ID must embed the ID near the end
+of the source, ideally on the last line, in the format `//# debugId=<DEBUG_ID>`:
+
+```javascript
+//# debugId=85314830-023f-4cf1-a267-535f4e37bb17
+```
+
+If the special `//# sourceMappingURL=` comment already exists in the file, it is
+recommended to place the `debugId` comment in the line above to maintain
+compatibility with existing tools.  Because the last line already has meaning in
+the existing specification for the `sourceMappingURL` comment, tools are
+required to examine the last 5 lines to discover the Debug ID.
+
+## JavaScript API for Debug ID Resolution
+
+Today `error.stack` in most runtimes only returns the URLs of the files referenced
```

Comment:
`Error.stack` returns an opaque string. With some luck you are able to parse a URL out of it though.